### PR TITLE
Improved filtering

### PR DESCRIPTION
**Pull request by alejoe91**
- moved out the butter filter creation in the init
- cache option now filters all the data and dumps it into a bin recording

Should fix:
https://github.com/SpikeInterface/spiketoolkit/issues/249  https://github.com/SpikeInterface/spiketoolkit/issues/248

I tested the performance on 120s 16 channels:

**Default filter**
```
rec_f = st.preprocessing.bandpass_filter(rec)

# filter all traces
t_start = time.time()
rec_f.get_traces()
print(time.time() - t_start)

2.017151117324829 s

# get waveform (before internal caching)
t_start = time.time()
wf1 = st.postprocessing.get_unit_waveforms(rec_f, sort)
print(time.time() - t_start)

2.023608446121216 s
```

**Cache** - the filtering is performed at initialization and dumped to file (a bit slower)
```
# initialization + filter + save
t_start = time.time()
rec_c = st.preprocessing.bandpass_filter(rec, cache=True)
print(time.time() - t_start)

3.8165152072906494 s

# get waveform 
t_start = time.time()
wf1 = st.postprocessing.get_unit_waveforms(rec_c, sort)
print(time.time() - t_start)

0.08617234230041504 s
```

@samuelgarcia @teristam @magland 
I think this will speed up **A LOT** the waveform retrieval. If you agree to this design I will propagate the cache option to all other preprocessors.

### CODE REVIEW

**Comment by colehurwitz** [comment]
Wow!! This is awesome

**Comment by colehurwitz** [comment]
Can this be implemented in the filter base class or does it have to be for each filter specifically?

**Comment by samuelgarcia** [comment]
@teristam : Yes there is a **delete_output_folder** option False by default.

**Comment by colehurwitz** [comment]
> @teristam : Yes there is a **delete_output_folder** option False by default.

I think that @teristam is referring to deleting the cached file when filtering not for when sorting.

**Comment by samuelgarcia** [comment]
Sorry.
I am just reading the code.
I was refering to another question.

@alejoe91 I am not sure to be happy that the trick of cache is done in the function and not in the classs itself.
Furthermore it could be done at a baseclass level like a mixin class **ProcessorCache**.

I will read it more carfully later.

**Comment by alejoe91** [comment]
I agree. This was just an example to test performance. @magland suggested an extractor called Cache recording extractor that can be initialized by passing a recording extractor. I think it's a good idea

**Comment by alejoe91** [comment]
@samuelgarcia I think that the function calls will still be the right place to wrap the preprocessor extractors in the cache recording object. I'll give it a try!

**Comment by alejoe91** [comment]
@teristam tested now on 7000 waveforms...

Without caching: 96 s (without running a `recording.get_traces()` before, as one would normally do)
With new caching: 0.67 s

**Comment by alejoe91** [comment]
When https://github.com/SpikeInterface/spikeextractors/pull/264 is merged tests should pass!

Let me know what you guys think!

**Comment by teristam** [comment]
@alejoe91 that's awesome! Just curious how would a disk-cached signal be used in the various sorters? Will the entire signal be loaded into memory?

File: spiketoolkit/preprocessing/bandpass_filter.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
 from .filterrecording import FilterRecording
 import numpy as np
 from scipy import special
+import tempfile
```
Comment (samuelgarcia) [inline spiketoolkit/preprocessing/bandpass_filter.py]: not needed anymore

**Comment by samuelgarcia** [comment]
Hi Alessio,
this looks good some questions:
  * why for FFT case the kernel is computed for each chunk ?
  * what append if we do this:
```python

rec0 = se.SomeExtrcator
rec_cache = bandpass_filter(rec0, freq_min=3000, freq_max=6000,
    type='butter', order=3, cache=True)
some_trace = rec_cache.get_traces(0, 1000)
del rec_cache
```
My guess is that the del will fail because some_trace still memmap the file that cannot be removed.

What do you think ?

**Comment by alejoe91** [comment]
> @alejoe91 that's awesome! Just curious how would a disk-cached signal be used in the various sorters? Will the entire signal be loaded into memory?
> 
> A typical workflow using the package will be
> 
> 1. Load the experimental data from disk
> 2. Make a RecordingExtractor
> 3. Do filtering
> 4. Do sorting
> 5. Calculate some metrics for curation
> 6. Save the sorting results
> 
> I am wondering if we save the filtered signals to disk, but later load it back to memory in sorting, does it offer any benefit than just keeping the entire filtered signal in memory?

@teristam 1. and 2. are the same: loading a recording creates a RecordingExtractor. 
Usually CLI-based sorters (klusta and SC) and matlab based (kilosort, ironclust, waveclus) make another copy of the signals in the output folder. However MS4 will benefit a lot from caching as it uses the recrding extractor directly!

**Comment by alejoe91** [comment]
@samuelgarcia @colehurwitz I think this is ready to merge on my side

**Comment by colehurwitz** [comment]
As long as @samuelgarcia is ok with it, I am ok with it :D

**Comment by samuelgarcia** [comment]
@alejoe91 : you did not answer to my comment from yesterday but it is totally OK for me.
:)

**Comment by alejoe91** [comment]
> Hi Alessio,
> this looks good some questions:
> 
> * why for FFT case the kernel is computed for each chunk ?
> * what append if we do this:
> 
> ```python
> rec0 = se.SomeExtrcator
> rec_cache = bandpass_filter(rec0, freq_min=3000, freq_max=6000,
>     type='butter', order=3, cache=True)
> some_trace = rec_cache.get_traces(0, 1000)
> del rec_cache
> ```
> 
> My guess is that the del will fail because some_trace still memmap the file that cannot be removed.
> 
> What do you think ?

Saw it now! I can test it tomorrow. Let's keep this open for the moment?

**Comment by alejoe91** [comment]
@samuelgarcia the kernel depends on the chunk size and that could change depending on padding. If you filter a range that starts from zero or at the end you will end up with a different kernel.

I tried what you suggested and it works with no problems. So maybe the traces are not memmaped because of this:
https://github.com/SpikeInterface/spikeextractors/blob/master/spikeextractors/extraction_tools.py#L245
The destructor is called and the file is deleted...

I also tried to use `np.memmap.transpose` instead and it still works. Strange. Should I merge anyways? I don't think users would delete the object and continue to do stuff.

**Comment by samuelgarcia** [comment]
So lets merge.
Stange that there is not bug...
Normally caching with memmap can lead lead to boudaries effect.
I guess under window it will be differents.
We will se later if we have problems, this cache is indeed a very good thing and the implementation is super simple, I like it.

**Pull request merged**